ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
+Also, a language that has a good expressive power already should not need a way to hack around with arbitrary code injection, like C macros do.
+
+| unstructured | structured |
+|--------------|------------|
+| (+) much easier to implement: just look for the `#word` symbols | (-) harder to implement: requires a full parsing |
+| (+) often language-agnostic (see the [C preprocessor](https://en.wikipedia.org/wiki/C_preprocessor). Familiar to C developers. | (-) a new syntax needs to be taught, not always self-explanatory |
+| (-) poorly integrated in the language, harder to read by humans | (+) is a well-designed syntax feature of the language |
+| (+) technically more expressive (e.g. [manipulating identifiers](https://en.wikipedia.org/wiki/C_preprocessor#Token_concatenation)) | (-) can only conditionaly include parts of the syntax tree, poor text generation capabilities. |
+| (-) behaves unpredictably, intent and behavior is hidden | (+) intent and behavior is visible at the usage site |
+| (-) no type checking, no syntax checking, because code is generated dynamically | (+) can statically check that conditional variants lead to syntactically correct and type-valid code |
+| (-) poor IDE support | (+) IDE can check all possible code paths |
+
+## Proposal
+
+We propose to leverage the [*attribute* syntax](https://www.w3.org/TR/WGSL/#attributes) to decorate syntax nodes with conditional compilation attributes.
+The compiler is provided with a set of '*compile-time features*' to enable (see below).
+
+### Definitions
+
+ * **Compile-time expression**: A *compile-time expression* is evaluated by the compiler before the rest of the program. It lives in the *compile-time scope*.
+
+ * **Compile-time scope**: The *compile-time scope* this is an independent scope from the [*module scope*](https://www.w3.org/TR/WGSL/#module-scope), meaning it cannot see any declarations from the source code, and its identifers are independent.
+
+ * **Compile-time feature**: A *compile-time feature* is an identifier that evaluates to a boolean. It is set to `true` if the feature is *enabled* during the compilation phase.
+
+ * **Compile-time attribute**: A *compile-time attribute* is parametrized by a *compile-time expression*. It is eliminated after the compilation phase but can affect the syntax node it decorates.
+
+### Location of *Compile-time attributes*
+
+(TODO: this needs more reflexion)
+
+A Compile-time attribute CAN decorate the following syntax nodes:
+ * structure declarations
+ * structure members
+ * function declarations
+ * ?? function formal parameter declarations ?? (is this needed?)
+ * variable declarations and value declarations
+ * all statements, except those disallowed below
+ * directives
+
+Because eliminating the decorated node would lead to invalid code, a Compile-time attribute CANNOT decorate the following syntax nodes:
```

Comment:
projects are doing some of these things with homegrown ifdef implementations

do we have an alternate solution for them? e.g. perhaps our other features may cover for their needs? or perhaps you've got a runtime condition mechanism in mind?

Reply:
I don't fully understand why decorating all of these nodes would lead to invalid code?

As in, this could work. It'd just eliminate the entire if-else statement
```
@if(feature_1)
if (a<b) {
...
} else {
...
}
```
Meanwhile that could reasonably be disallowed
```
if (a<b) {
...
} 
@if(feature_1)
else {
...
}
```
Because people are supposed to write
```
if (a<b) {
...
} else {
@if(feature_1) {
...
}
}
```

Reply:
What are you referring to?

Reply:
> Because eliminating the decorated node would lead to invalid code, a Compile-time attribute CANNOT decorate the following syntax nodes:
>  * bodies of function declarations, switch statements, switch clauses, loop statements, for statements, while statements, if/else statements
> * function return types

Reply:
Is that just because of how the WGSL grammar is set up?

Reply:
I will clarify this. But I refered to the **bodies**, because the grammar for these is like this:
`attribute* if/else/while... () CompoundStatement`, where `compoundStatent` is the body and looks like `attribute* { Statement* }`

so you don't want to have
`while (true) @if(false) {}`

because the CompoundStatement is not optional.

Reply:
> Meanwhile that could reasonably be disallowed
> 
> ```
> if (a<b) {
> ...
> } 
> @if(feature_1)
> else {
> ...
> }
> ```
> 
> Because people are supposed to write
> 
> ```
> if (a<b) {
> ...
> } else {
> @if(feature_1) {
> ...
> }
> }
> ```

Why? your first variant looks more correct to me. Should you not be able to erase an `else`?

Reply:
I see, I read it as 

> (bodies of function declarations), switch statements, switch clauses, loop statements, for statements, while statements, if/else statements

instead of the intended

> bodies of (function declarations, switch statements, switch clauses, loop statements, for statements, while statements, if/else statements)

> Should you not be able to erase an `else`?

The WGSL grammar doesn't allow an attribute in that position. Rust also disallows attributes in that position.

Reply:
> The WGSL grammar doesn't allow an attribute in that position. Rust also disallows attributes in that position.

Oh I missed that one, thanks! about rust: in rust if/else can be an expression whereas if alone cannot. In wgsl, this is not a problem as if/else cannot be and expression. So I suggest to add else to the list of places where we would allow attributes.